classes/orcid/actions/VerifyAuthorWithOrcid.php

```diff
@@ -0,0 +1,151 @@
+ */
+
+namespace PKP\orcid\actions;
+
+use APP\author\Author;
+use APP\core\Application;
+use APP\core\Request;
+use APP\facades\Repo;
+use APP\orcid\actions\SendSubmissionToOrcid;
+use APP\template\TemplateManager;
+use Carbon\Carbon;
+use GuzzleHttp\Exception\GuzzleException;
+use PKP\orcid\OrcidManager;
+use PKP\submission\PKPSubmission;
+
+class VerifyAuthorWithOrcid
+{
+    public function __construct(
+        private Author $author,
+        private Request $request,
+        private array $templateVarsToSet = []
+    ) {
+    }
+
+    public function execute(): self
+    {
+        $context = $this->request->getContext();
+
+        // Fetch the access token
+        $oauthTokenUrl = OrcidManager::getApiPath($context) . OrcidManager::OAUTH_TOKEN_URL;
+
+        $httpClient = Application::get()->getHttpClient();
+        $headers = ['Accept' => 'application/json'];
+        $postData = [
+            'code' => $this->request->getUserVar('code'),
+            'grant_type' => 'authorization_code',
+            'client_id' => OrcidManager::getClientId($context),
+            'client_secret' => OrcidManager::getClientSecret($context)
+        ];
+
+        OrcidManager::logInfo('POST ' . $oauthTokenUrl);
+        OrcidManager::logInfo('Request headers: ' . var_export($headers, true));
+        OrcidManager::logInfo('Request body: ' . http_build_query($postData));
+
+        try {
+            $response = $httpClient->request(
+                'POST',
+                $oauthTokenUrl,
+                [
+                    'headers' => $headers,
+                    'form_params' => $postData,
+                    'allow_redirects' => ['strict' => true],
+                ],
+            );
+
+            if ($response->getStatusCode() !== 200) {
+                OrcidManager::logError('VerifyAuthorWithOrcid::execute - unexpected response: ' . $response->getStatusCode());
+                $this->addTemplateVar('authFailure', true);
+            }
+            $results = json_decode($response->getBody(), true);
+
+            // Check for errors
+            OrcidManager::logInfo('Response body: ' . print_r($results, true));
+            if (($results['error'] ?? null) === 'invalid_grant') {
+                OrcidManager::logError('Authorization code invalid, maybe already used');
+                $this->addTemplateVar('authFailure', true);
+            }
+            if (isset($results['error'])) {
+                OrcidManager::logError('Invalid ORCID response: ' . $results['error']);
+                $this->addTemplateVar('authFailure', true);
+            }
+
+            // Check for duplicate ORCID for author
+            $orcidUri = OrcidManager::getOrcidUrl($context) . $results['orcid'];
+            if (!empty($this->author->getOrcid()) && $orcidUri !== $this->author->getOrcid()) {
+                $this->addTemplateVar('duplicateOrcid', true);
```

Comment:
I think the difference may come from the source of the request, whether it's initiated by a user in their user profile or as part of the submission workflow process. When it's initiated by the user in their profile directly, the ORCID is associated with a user, but when it's initiated as part of the submission workflow, it's only associated with the author attached to that submission.

I think this is where the language is coming from (it's stored for the author for this submission rather than being stored for a user). That being said, I don't know that language that specific is necessarily helpful.

`orcid.verify.duplicateOrcid` makes reference to a submission, which is confusing while `orcid.verify.duplicateOrcidAuthor` refers to multiple ORCIDs/contributors (`Duplicate ORCiDs for contributors detected.`).

I think short-term we should probably keep it as-is, but maybe revisit the language used. I don't think its necessarily clear a person when they click the link to verify their ORCID in an email that it will be associated with one submission specifically. What do you think?

Reply:
Yes, that sounds good...